Mission 2

forestream opened the pull request:
- `createRoot`, `render` 함수 제작
- `VirtualNode`, `VirtualDOM` 클래스 제작
- `useState` 제작

- HTMLElement 컴포넌트는 문제 없는데 함수 컴포넌트에 children을 전달하면 오류가 뜨는데 어떻게 해결하는지 모르겠습니다.
![image](https://github.com/user-attachments/assets/25a707ce-69a4-47ec-b4b3-8f8aab9ed128)
`<p>foo</p>` 는 오류가 안 생기는데
`<Component>bar</Component>` 는 자꾸 오류가 뜨네요.

- 이벤트 핸들러를 처리해야 하는데 `input` 이벤트에서 `setState`가 호출되면 리렌더 후 인풋 요소에서 포커스가 사라져서 한 글자 입력하고 포커스하고 입력하고 포커스하고 해야 하는 문제를 어떻게 할지 고민 중입니다.
현재 코드는 사용자 입력 시 `change` 이벤트핸들러를 사용하고 있습니다.

forestream committed:
createRoot, render 함수 개발

forestream committed:
Virtual DOM 트리에서 real DOM 생성 함수 개발

forestream committed:
useState 개발

forestream committed:
onClick, onChange 프롭스 처리

- 개선 필요

forestream committed:
createElement 인수에 chlidren이 없을 경우 children: [] 을
children: undefined 로 반환하도록 수정

- 전자의 경우 컴포넌트에 children을 전달하지 않아도 children 프롭에 빈 배열이 있는 것처럼 인식했습니다.

YOOJS1205 commented on src/react/jsx-runtime/index.ts:
```diff
@@ -3,20 +3,63 @@ import {
-	props: PropsWithoutChildren,
-	...children: Children
-): ReactElement {
+const React = (function () {
```
즉시 실행 함수를 통해 참조하는 변수들을 잘 보호해주셨습니다!

YOOJS1205 commented on src/react/jsx-runtime/index.ts:
```diff
@@ -3,20 +3,63 @@ import {
+		};
+	}
+
+	function useState<T>(initState: T): [T, (newState: T) => void] {
```
전반적으로 `useState`의 로직은 잘 작성해주셨습니다!

다만, 현재 구조라면 이전 상태와 현재 상태의 값이 바뀌더라도 `render` 함수가 호출될 것 같습니다.

따라서, 이전 상태와 현재 상태를 비교해서 `render` 함수를 호출하지 않는 로직을 추가해주시면 좋을 것 같아요!

구현하실 때 **얕은 비교 (shallowEqual)** 키워드를 참조해서 적용해보세요!

YOOJS1205 commented on src/react/jsx-runtime/index.ts:
```diff
@@ -3,20 +3,63 @@ import {
+			states.push(initState);
+		}
+
+		function setState(newState: T) {
```
`setState`가 함수형 업데이트도 가능하도록 지원해보면 어떨까요? React에서 함수형 업데이트는 상태 업데이트에서 동기성 보장을에 핵심적인 역할을 합니다.

```ts
setCount((prev) => prev + 1);
```

만약 작성하신 로직에서 아래의 함수를 호출하면 `count`의 값은 어떻게 될까요?

```ts
const handleClickPlusButton = () => {
  setCount(count + 1);
  setCount(count + 1);
  setCount(count + 1);
}
```

적용하실 때 **함수형 업데이트, setState의 비동기적 특성** 키워드를 참고해보세요!

YOOJS1205 reviewed (commented):
전반적으로 잘 구성해주셨습니다!

우선 현재 onInput, onChange 관련해서 마주하신 문제는 3주차 과제에서 diffing 로직을 구현하시면 자연스럽게 해결이 됩니다.

HTML input 태그에서 `onInput`, `onChange` 이벤트의 차이와 React에서는 두 이벤트를 어떻게 처리하는지 공부해보시면 좋을 것 같아요!

캡쳐해주신 문제는 제가 확인했을 땐 재현이 안되어서, 재현할 수 있는 코드를 제공해주시면 추가로 확인해볼게요!

고생하셨습니다.

forestream committed:
rerender 함수 개발

- setState 호출 후 기존 VirtualDOM을 순회하며 각 노드의 states가 업데이트 됐는지 확인합니다.
상태가 업데이트 된 컴포넌트가 있으면 해당 노드부터 VirtualDOM 트리를 만들어서 기존 트리의 노드와 교체하고
실제 DOM에 적용합니다.

forestream committed:
투두리스트 기본 디자인 형성

forestream committed:
실제 DOM 생성 시 HTMLElement attribute 할당 방식 수정

- 프롭 키를 할당 연산자를 통하여 할당할 수 있는지 판단하는 hasSetter 유틸 함수 개발
ex: input.value = 'foo';

- hasSetter(): true 인 모든 프롭 키는 프롭 값을 할당하도록 했습니다.

forestream committed:
DOM 생성 메서드 수정

- 기존 ReactElement를 인수로 받던 realizeVirtualDOMTree를 인수 없이 호출하도록 변경.
호출하기 전에 generateVirtualDOMTree >> virtualDom.root.appendChild로 트리를 미리 형성해두고, 그 상태를 기반으로 호출하는 것이 함수 기능 분리 측면에서 좋아 보임.

forestream committed:
실제 DOM 비교 로직 개발

- 변경이 필요한 요소만 업데이트 할 수 있도록 VirtualDOM.compare 메서드 개발
- realizeVirtualDOMTree 실행 중 컴포넌트 가상 노드를 HTMLTemplateElement로 실제 노드 생성하던 방식에서, 아예 실제 노드를 생성하거나 삭제하지 않는 방식으로 변경
- realizeVirtualDOMTree가 리렌더링 때 쓰일 수 있도록 수정.
realChildNodeIndex 변수를 만들어서 하위 노드에서 한 단계 올라올 때 이전 인덱스를 기억할 수 있도록 함.
ex) ul 아래 li 요소가 3개 있고 첫번째 li 요소 순회를 마치면 두번째 li 요소를 기존 요소와 비교할 수 있어야 함.

forestream committed:
useState 인수 타입 확장

- 함수로 상태값 생성할 수 있도록 수정
